### PR TITLE
Issue #237: Handle SIGTERM and gracefully terminate

### PR DESCRIPTION
**Pull request by kevin-bates**
Add signal handler for termination requests and invoke stop sequence to shutdown kernels and stop application.

### CODE REVIEW

**Comment by parente** [comment]
It looks like travis might be failing because of a backwards incompatibility with a second order dependency. The `ipykernel_launcher` appearing in the test log output is new.

I did confirm that notebook 4.x is still being installed, as desired for the moment. @minrk do you have any idea what else might have rev'ed and broken?

File: kernel_gateway/gatewayapp.py [hunk 1]
```diff
@@ -528,5 +530,8 @@ def shutdown(self):
         """Stop all kernels in the pool."""
         self.personality.shutdown()
 
+    def _signal_stop(self, sig, frame):
+        self.log.info("Received signal to terminate.")
+        ioloop.IOLoop.current().stop()
```
Comment (rolweber) [inline kernel_gateway/gatewayapp.py]: All other places use `self.io_loop`, it seems. But if that is used here, the signal handler should be installed after assigning `self.io_loop`.

**Comment by minrk** [comment]
> @minrk do you have any idea what else might have rev'ed and broken?

There is a kernelspec installed where the executable is only `python` (this happens from wheels of ipykernel 4.6). This relies on resolution via PATH *in the subprocess*. When the env is overridden in KG, PATH is unset and thus the system default is applied. The result is that the kernel is launched with `/usr/bin/python` instead of the env's Python.

The fix should be for rest-api env to override the default env, not replace it. #239 should fix this.

File: kernel_gateway/gatewayapp.py (hunk 1 above)
Comment (minrk) [inline kernel_gateway/gatewayapp.py]: self.io_loop makes sense. `IOLoop.current()` uses thread-local storage, so may not find the right IO Loop, depending on whether the application and the event handler are running in the same thread.

**Comment by parente** [comment]
> There is a kernelspec installed where the executable is only python (this happens from wheels of ipykernel 4.6).

Noting this for future reference mainly:

https://github.com/ipython/ipykernel/commit/654c7829c9b7e6984a0689d2b47c364a623436a1#diff-2eeaed663bd0d25b7e608891384b7298R93 appears to be the PR and line that sets the python executable path upon wheel install.

File: kernel_gateway/gatewayapp.py (hunk 1 above)
Comment (kevin-bates) [inline kernel_gateway/gatewayapp.py]: Excellent feedback - thank you for catching this.  I've tested the recommended changes and posted the commit.

**Comment by parente** [comment]
Thanks @kevin-bates . @rolweber feel free to merge if everything looks good to you in the future, unless you want me to have a look too.

**Pull request merged**